### PR TITLE
Adding os_quota support to the OpenStack modules

### PR DESCRIPTION
**Pull request by mgale**
##### ISSUE TYPE
- New Module Pull Request
##### COMPONENT NAME

cloud/openstack/os_quota
##### ANSIBLE VERSION

```
ansible 2.0.2.0
```
##### SUMMARY

Allow full control over setting and updating project quotas in OpenStack.

Does require shade > 1.9.0
Tested with shade 1.9.1

```
Example Playbook:

---

- include_vars: /etc/openstack_deploy/projects.yml
  tags: ['always']
  ignore_errors: yes

- name: Update srt_stress_testing quota
  os_quota:
    name: "{{ item.name }}"
    backup_gigabytes: "{{ item.backup_gigabytes }}"
    backups: "{{ item.backups }}"
    cores: "{{ item.cores }}"
    fixed_ips: "{{ item.fixed_ips }}"
    floating_ips: "{{ item.floating_ips }}"
    gigabytes: "{{ item.gigabytes }}"
    injected_file_size: "{{ item.injected_file_size }}"
    injected_files: "{{ item.injected_files }}"
    injected_path_size: "{{ item.injected_path_size }}"
    instances: "{{ item.instances }}"
    port: "{{ item.port }}"
    key_pairs: "{{ item.key_pairs }}"
    per_volume_gigabytes: "{{ item.per_volume_gigabytes }}"
    properties: "{{ item.properties }}"
    ram: "{{ item.ram }}"
    secgroup_rules: "{{ item.secgroup_rules }}"
    secgroups: "{{ item.secgroups }}"
    server_group_members: "{{ item.server_group_members }}"
    server_groups: "{{ item.server_groups }}"
    snapshots: "{{ item.snapshots }}"
    volumes: "{{ item.volumes }}"
    volumes_types:
      volumes_lvm: "{{ item.volumes_lvm }}"
    snapshots_types:
      snapshots_lvm: "{{ item.snapshots_lvm }}"
    gigabytes_types:
      gigabytes_lvm: "{{ item.gigabytes_lvm }}"
  with_items:
    - "{{ projects }}"
```

### CODE REVIEW

**Comment by gregdek** [comment]
Thanks @mgale. To the current maintainers, @emonty, @shrews, @juliakreger, @j2sol, @rcarrillocruz please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with text 'shipit', 'needs_revision' or 'close_me' as appropriate.

[This message brought to you by your friendly Ansibull-bot.]

File: cloud/openstack/os_quota.py
Comment (omgjlk) [inline cloud/openstack/os_quota.py]: This appears to be missing some data.

File: cloud/openstack/os_quota.py
Comment (omgjlk) [inline cloud/openstack/os_quota.py:286]: There is some odd whitespacing of new lines in each of these functions.

File: cloud/openstack/os_quota.py
Comment (omgjlk) [inline cloud/openstack/os_quota.py]: These should be removed.

File: cloud/openstack/os_quota.py
Comment (omgjlk) [inline cloud/openstack/os_quota.py]: This should be moved to a function like the other os_ modules use to detect if change is needed.

File: cloud/openstack/os_quota.py
Comment (omgjlk) [inline cloud/openstack/os_quota.py]: This seems like a weird thing to turn into a function you call exactly once. Why make it a function?

**Comment by omgjlk** [comment]
Whitespace concerns, comments in code, and this module doesn't support check mode.

needs_revision

**Comment by gregdek** [comment]
Thanks @mgale for this PR. This PR requires revisions, either because it fails to build or by reviewer request. Please make the suggested revisions. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by emonty** [comment]
@mgale this is great in general (although I agree with @j2sol on his review) Once we land this, I'd like to move a chunk of logic you have in the module into the shade library. Is that ok with you? shade is apache2 licensed, so I'll need your permission to do so.

**Comment by mgale** [comment]
@emonty @j2sol thanks so much for the feedback, I will try and get those changes completed asap, I am on vacation and without network access from today until Sept 1. I really do appreciate the time and effort of the code reviews.

@emonty you have my permission to use the code however you see fit, inside the shade lib or anywhere else.

**Comment by mgale** [comment]
ready_for_review

**Comment by gregdek** [comment]
Thanks @mgale. To the current maintainers, @emonty, @shrews, @juliakreger, @j2sol, @rcarrillocruz please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with text 'shipit', 'needs_revision' or 'close_me' as appropriate.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by mgale** [comment]
@gregdek Hey, am I just waiting for code reviews at this point? Is anyone waiting on me for anything?

**Comment by omgjlk** [comment]
Yes, you're waiting for reviews.

File: cloud/openstack/os_quota.py
Comment (omgjlk) [inline cloud/openstack/os_quota.py]: you're returning this as ansible_facts, which will set all of these things as host vars. You don't document that in the documentation, and it doesn't match the return spec you do document. Is there a reason you're turning these into facts? Either way the docs should line up with the code or the code with the docs.

**Review by omgjlk — CHANGES_REQUESTED** [review]
needs_revision

File: cloud/openstack/os_quota.py
Comment (mgale) [reply cloud/openstack/os_quota.py]: I am actually not sure what is the correct practice here, I ended up doing it this way based on another example. Can I just remove the ansible_facts part and return the openstack_quotas var?

If so I will update the documentation and code?

Comment (omgjlk) [reply cloud/openstack/os_quota.py]: I would recommend what you suggest. Remove `ansible_facts` and just have `openstack_quotas`.  That should match the written documentation.

**Comment by mgale** [comment]
ready_for_review

File: cloud/openstack/os_quota.py
Comment (mgale) [reply cloud/openstack/os_quota.py]: Done.

**Comment by mgale** [comment]
Hello @j2sol , any chance you can take another look?

File: cloud/openstack/os_quota.py
Comment (omgjlk) [inline cloud/openstack/os_quota.py]: This will need to say 2.3, I'm sorry that it missed the 2.2 cut off :(

File: cloud/openstack/os_quota.py
Comment (omgjlk) [inline cloud/openstack/os_quota.py:190]: The comment says List, but the action seems to be setting cores, what is intended here?

**Review by omgjlk — CHANGES_REQUESTED** [review]
Code mostly looks good, but a couple small things to change regarding the comments. I haven't tested the code personally  though.

**Comment by omgjlk** [comment]
needs_revision

**Comment by mgale** [comment]
ready_for_review

**Comment by mercantiandrea** [comment]
I didn't know this pull so I have done a new pull request for quotas update in OpenStack.
https://github.com/ansible/ansible-modules-extras/pull/3414
I think that this is a better work than mine. 

There is only one functionality that is present this pull request but it isn't present in the other pull request:
the restore of the default quotas with the parameter "state: absent" and the call "cloud.delete_volume_quotas(module.params['project'])".
Could you please add this functionality?

Finally I would change parameter "name" with "project".

**Comment by mercantiandrea** [comment]
@j2sol I have tested personally the code and it works well.

**Comment by mgale** [comment]
@mercantiandrea I will look into making those changes.

**Comment by mgale** [comment]
ready_for_review

**Comment by mgale** [comment]
@mercantiandrea Can you look this over and see if it is what you expected? Also for the current maintainers, @emonty, @shrews, @juliakreger, @j2sol, @rcarrillocruz. I believe I ran into a shade bug when calling delete_network_quotas when a custom quota has not yet been created. I worked around the issue however I am not sure I handled it properly. You can see the workaround on lines 443-448.

**Pull request closed without merging**
